### PR TITLE
Support constant lr with cooldown

### LINKED ISSUE #35449: Support Constant Learning Rate with Cooldown
### Feature request

In `transformers.optimization` support `constant learning rate with cooldown` functions.

### Motivation

This method will implement that scaling experiments can be performed with significantly reduced compute and GPU hours by utilizing fewer but reusable training runs.
[SmolLM](https://huggingface.co/blog/smollm) had used this method to train a series of SOTA small language models.
Paper: [https://arxiv.org/pdf/2405.18392](https://arxiv.org/pdf/2405.18392)

### Your contribution

I've created a branch, I'm finishing the implementation of these functions, and intend to submit a PR.

### PR DESCRIPTION
**Pull request by LoserCheems**
# What does this PR do?

Fixes #35449
Added the 'warmup_stable_cooldown' learning rate scheduler method.
This method allows three phases, linear warmup, stable, and cooldown, where cooldown can be done using the `linear`, `cosine` and `1-sqrt` methods.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@muellerzr and @SunMarc

### CODE REVIEW

**Comment by SunMarc** [comment]
Thanks for the PR @LoserCheems, we already have warmup_stable_decay scheduler with `get_wsd_schedule` that does practically the same thing as the new scheduler you proposed. I think what you be good is to maybe add more options to the decay stage of the wsd_scheduler as you did in this PR (e.g linear/ 1-sqrt) WDYT ?

**Comment by LoserCheems** [comment]
> Thanks for the PR @LoserCheems, we already have warmup_stable_decay scheduler with `get_wsd_schedule` that does practically the same thing as the new scheduler you proposed. I think what you be good is to maybe add more options to the decay stage of the wsd_scheduler as you did in this PR (e.g linear/ 1-sqrt) WDYT ?

oh, thank for your suggestion, I intend to integrate different cooldown methods and minimum learning rate into `get_wsd_schedule`.
In addition, is it possible to rename `get_wsd_schedule` to `get_wsc_schedule, which is more in line with the three phases of warmup, stable, cooldown.

**Comment by SunMarc** [comment]
Thanks ! We can't really rename it wsc, wsd is a real term that appears in the MiniCPM paper.

**Comment by LoserCheems** [comment]
> Thanks ! We can't really rename it wsc, wsd is a real term that appears in the MiniCPM paper.

Thank you, the renaming of the function is done.

File: src/transformers/optimization.py [hunk 1]
```diff
@@ -391,47 +391,75 @@ def _get_wsd_scheduler_lambda(
...
+        return max(0.0, factor)
+
     return min_lr_ratio
 
 
 def get_wsd_schedule(
     optimizer: Optimizer,
     num_warmup_steps: int,
-    num_stable_steps: int,
+    num_training_steps: int,
```
Comment (SunMarc) [inline src/transformers/optimization.py:436]: We can't do that without a deprecating cycle as this is breaking. Another idea would be to make both num_stable_steps and num_training_steps optional then create a check that enforce that one of them must be specified. Otherwise, can you try to keep `num_stable_steps` instead ? Is there a specific reason why you want to use num_training_steps ?

File: src/transformers/optimization.py [hunk 2]
```diff
@@ -391,47 +391,75 @@ def _get_wsd_scheduler_lambda(
     current_step: int,
     *,
     num_warmup_steps: int,
-    num_stable_steps: int,
+    num_training_steps: int,
```
Comment (SunMarc) [inline src/transformers/optimization.py:394]: as this is private, it is fine to have breaking changes here

**Review by SunMarc — COMMENTED** [review]
Thanks for the PR ! Left a comment

File: src/transformers/optimization.py (hunk 1 above)
Comment (LoserCheems) [reply src/transformers/optimization.py:436]: Because I noticed that all the other functions use `num_training_steps` and only this one uses `num_stable_steps`, I changed the logic of this function in `get_scheduler`, We allow `num_warmup_steps` and `num_training_steps` to be passed normally, just like the most commonly used `get_cosine_with_min_lr_schedule_with_warmup`.

Comment (LoserCheems) [reply src/transformers/optimization.py:436]: I can write an extra private function to add a branch to `get_scheduler`, which will be updated shortly.

File: src/transformers/optimization.py [hunk 3]
```diff
@@ -540,9 +599,6 @@ def scheduler_hook(param):
     if name == SchedulerType.INVERSE_SQRT:
         return schedule_func(optimizer, num_warmup_steps=num_warmup_steps)
 
-    if name == SchedulerType.WARMUP_STABLE_DECAY:
-        return schedule_func(optimizer, num_warmup_steps=num_warmup_steps, **scheduler_specific_kwargs)
-
     # All other schedulers require `num_training_steps`
     if num_training_steps is None:
         raise ValueError(f"{name} requires `num_training_steps`, please provide that argument.")
```
Comment (SunMarc) [inline src/transformers/optimization.py:604]: because of this, we will always trigger an error asking the user to pass num_training_steps. This is a breaking change,

Comment (LoserCheems) [reply src/transformers/optimization.py]: Thank you, I've moved the wsd before the decision.

File: src/transformers/optimization.py [hunk 4]
```diff
@@ -391,47 +391,96 @@ def _get_wsd_scheduler_lambda(
...
+                factor = 1.0 - math.sqrt(progress)
+            else:
+                raise ValueError(f"Unknown decay type: {decay_type}, expected 'linear', 'cosine' or '1-sqrt'")
+            factor = factor * (1.0 - min_lr_ratio) + min_lr_ratio
+            return max(0.0, factor)
+    else:
+        if current_step < num_warmup_steps + num_stable_steps:
+            return 1.0
+
+        if current_step < num_warmup_steps + num_stable_steps + num_decay_steps:
```
Comment (SunMarc) [inline src/transformers/optimization.py:435]: Instead of allowing both `num_training_steps` and `num_stable_steps` in `_get_wsd_scheduler_lambda` , can we just keep one in this function, this will make the function easier to follow. Then in `get_wsd_schedule`, we can, for example, deduce `num_stable_steps` from `num_training_steps`.

Comment (LoserCheems) [reply src/transformers/optimization.py]: I defined a `stable_stage` to merge the two branches.

File: src/transformers/optimization.py [hunk 5]
```diff
@@ -391,47 +391,81 @@ def _get_wsd_scheduler_lambda(
     current_step: int,
     *,
     num_warmup_steps: int,
+    num_training_steps: int,
     num_stable_steps: int,
```
Comment (SunMarc) [inline src/transformers/optimization.py:395]: I still feel like we shouldn't need to pass both values. 
Can we just keep num_stable_steps and in `get_wsd_schedule`, we set 
`num_stable_steps = num_training_steps - num_decay_steps - num_warmup_steps` if `num_training_steps is not None` ?

Comment (LoserCheems) [reply src/transformers/optimization.py]: I move stable number to `get_wsd_schedule`.
The `Trainer` is passed `num_training_steps` instead of `num_stable_steps` by default, I think using `num_training_steps` reduces passing `lr_scheduler_kwargs`...

Comment (LoserCheems) [reply src/transformers/optimization.py]: If we use `max_steps` in `Trainer`, we have to manually calculate the `stable` ratio in order to pass `num_stable_steps`, and this calculation is more complicated if we use `num_train_epoces`...

**Review by SunMarc — COMMENTED** [review]
Nice ! Just a few nits and we should be able to merge this soon ! Thanks for being reactive

**Comment by LoserCheems** [comment]
> Nice ! Just a few nits and we should be able to merge this soon ! Thanks for being reactive

Everything is ready!

File: src/transformers/optimization.py [hunk 6]
```diff
@@ -393,45 +393,71 @@ def _get_wsd_scheduler_lambda(
...
 
 def get_wsd_schedule(
     optimizer: Optimizer,
     num_warmup_steps: int,
-    num_stable_steps: int,
+    num_training_steps: int,
     num_decay_steps: int,
+    num_stable_steps: int = None,
+    warmup_type: str = "linear",
+    decay_type: str = "linear",
```
Comment (SunMarc) [inline src/transformers/optimization.py:435]: It is `cosine` by default in the previous implementation

File: src/transformers/optimization.py [hunk 7]
```diff
@@ -393,45 +393,71 @@ def _get_wsd_scheduler_lambda(
...
+        factor = factor * (1.0 - min_lr_ratio) + min_lr_ratio
+        return max(0.0, factor)
     return min_lr_ratio
 
 
 def get_wsd_schedule(
     optimizer: Optimizer,
     num_warmup_steps: int,
-    num_stable_steps: int,
+    num_training_steps: int,
```
Comment (SunMarc) [inline src/transformers/optimization.py:431]: `num_training_steps` should be optional too

**Review by SunMarc — APPROVED** [review]
LGTM ! Thanks for iterating ! Just a few nits

**Comment by LoserCheems** [comment]
> LGTM ! Thanks for iterating ! Just a few nits
It's done!

**Review by SunMarc — APPROVED** [review]
LGTM!

**Comment by LoserCheems** [comment]
Thank you, this PR is ready for merge.

**Comment by SunMarc** [comment]
gentle ping @ArthurZucker

**Comment by LoserCheems** [comment]
Sorry @SunMarc, is there anything wrong with this PR?

**Review by SunMarc — APPROVED** [review]
last nit and i will merge the pr
